3.2.12在键盘空白处滑动时 程序崩溃

**Describe the bug**
在键盘空白处滑动时，程序崩溃

**To Reproduce**
Steps to reproduce the bug:
横屏模式，滑动键盘空白处。

**Expected behavior**
无响应，或者向左滑时为返回。

**Log**
```
E/InputEventReceiver(31947): Exception dispatching input event.
E/MessageQueue-JNI(31947): Exception in MessageQueue callback: handleReceiveCallback
E/MessageQueue-JNI(31947): java.lang.ArrayIndexOutOfBoundsException: length=33; index=-1
E/MessageQueue-JNI(31947): 	at com.osfans.trime.ime.keyboard.KeyboardView$1.onFling(KeyboardView.java:527)
E/MessageQueue-JNI(31947): 	at android.view.GestureDetector.onTouchEvent(GestureDetector.java:767)
E/MessageQueue-JNI(31947): 	at com.osfans.trime.ime.keyboard.KeyboardView.onModifiedTouchEvent(KeyboardView.java:1580)
E/MessageQueue-JNI(31947): 	at com.osfans.trime.ime.keyboard.KeyboardView.onTouchEvent(KeyboardView.java:1543)
E/MessageQueue-JNI(31947): 	at android.view.View.dispatchTouchEvent(View.java:15111)
E/MessageQueue-JNI(31947): 	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3170)
E/MessageQueue-JNI(31947): 	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2846)
E/MessageQueue-JNI(31947): 	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3170)
E/MessageQueue-JNI(31947): 	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2846)
E/MessageQueue-JNI(31947): 	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3170)
E/MessageQueue-JNI(31947): 	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2846)
E/MessageQueue-JNI(31947): 	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3170)
E/MessageQueue-JNI(31947): 	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2846)
E/MessageQueue-JNI(31947): 	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3170)
E/MessageQueue-JNI(31947): 	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2846)
E/MessageQueue-JNI(31947): 	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3170)
E/MessageQueue-JNI(31947): 	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2846)
E/MessageQueue-JNI(31947): 	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3170)
E/MessageQueue-JNI(31947): 	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2846)
E/MessageQueue-JNI(31947): 	at com.android.internal.policy.DecorView.superDispatchTouchEvent(DecorView.java:569)
E/MessageQueue-JNI(31947): 	at com.android.internal.policy.PhoneWindow.superDispatchTouchEvent(PhoneWindow.java:1918)
E/MessageQueue-JNI(31947): 	at android.app.Dialog.dispatchTouchEvent(Dialog.java:960)
E/MessageQueue-JNI(31947): 	at android.inputmethodservice.SoftInputWindow.dispatchTouchEvent(SoftInputWindow.java:151)
E/MessageQueue-JNI(31947): 	at com.android.internal.policy.DecorView.dispatchTouchEvent(DecorView.java:517)
E/MessageQueue-JNI(31947): 	at android.view.View.dispatchPointerEvent(View.java:15384)
E/MessageQueue-JNI(31947): 	at android.view.ViewRootImpl$ViewPostImeInputStage.processPointerEvent(ViewRootImpl.java:6907)
E/MessageQueue-JNI(31947): 	at android.view.ViewRootImpl$ViewPostImeInputStage.onProcess(ViewRootImpl.java:6678)
E/MessageQueue-JNI(31947): 	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:6125)
E/MessageQueue-JNI(31947): 	at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:6187)
E/MessageQueue-JNI(31947): 	at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:6148)
E/MessageQueue-JNI(31947): 	at android.view.ViewRootImpl$AsyncInputStage.forward(ViewRootImpl.java:6322)
E/MessageQueue-JNI(31947): 	at android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:6156)
E/MessageQueue-JNI(31947): 	at android.view.ViewRootImpl$AsyncInputStage.apply(ViewRootImpl.java:6379)
E/MessageQueue-JNI(31947): 	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:6129)
E/MessageQueue-JNI(31947): 	at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:6187)
E/MessageQueue-JNI(31947): 	at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:6148)
E/MessageQueue-JNI(31947): 	at android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:6156)
E/MessageQueue-JNI(31947): 	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:6129)
E/MessageQueue-JNI(31947): 	at android.view.ViewRootImpl.deliverInputEvent(ViewRootImpl.java:9288)
E/MessageQueue-JNI(31947): 	at android.view.ViewRootImpl.doProcessInputEvents(ViewRootImpl.java:9239)
E/MessageQueue-JNI(31947): 	at android.view.ViewRootImpl.enqueueInputEvent(ViewRootImpl.java:9188)
E/MessageQueue-JNI(31947): 	at android.view.ViewRootImpl$WindowInputEventReceiver.onInputEvent(ViewRootImpl.java:9439)
E/MessageQueue-JNI(31947): 	at android.view.InputEventReceiver.dispatchInputEvent(InputEventReceiver.java:276)
E/MessageQueue-JNI(31947): 	at android.os.MessageQueue.nativePollOnce(Native Method)
E/MessageQueue-JNI(31947): 	at android.os.MessageQueue.next(MessageQueue.java:339)
E/MessageQueue-JNI(31947): 	at android.os.Looper.loopOnce(Looper.java:186)
E/MessageQueue-JNI(31947): 	at android.os.Looper.loop(Looper.java:351)
E/MessageQueue-JNI(31947): 	at android.app.ActivityThread.main(ActivityThread.java:8381)
E/MessageQueue-JNI(31947): 	at java.lang.reflect.Method.invoke(Native Method)
E/MessageQueue-JNI(31947): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:584)
E/MessageQueue-JNI(31947): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1013)
D/AndroidRuntime(31947): Shutting down VM
--------- beginning of crash
E/AndroidRuntime(31947): FATAL EXCEPTION: main
E/AndroidRuntime(31947): Process: com.osfans.trime, PID: 31947
E/AndroidRuntime(31947): java.lang.ArrayIndexOutOfBoundsException: length=33; index=-1
E/AndroidRuntime(31947): 	at com.osfans.trime.ime.keyboard.KeyboardView$1.onFling(KeyboardView.java:527)
E/AndroidRuntime(31947): 	at android.view.GestureDetector.onTouchEvent(GestureDetector.java:767)
E/AndroidRuntime(31947): 	at com.osfans.trime.ime.keyboard.KeyboardView.onModifiedTouchEvent(KeyboardView.java:1580)
E/AndroidRuntime(31947): 	at com.osfans.trime.ime.keyboard.KeyboardView.onTouchEvent(KeyboardView.java:1543)
E/AndroidRuntime(31947): 	at android.view.View.dispatchTouchEvent(View.java:15111)
E/AndroidRuntime(31947): 	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3170)
E/AndroidRuntime(31947): 	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2846)
E/AndroidRuntime(31947): 	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3170)
E/AndroidRuntime(31947): 	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2846)
E/AndroidRuntime(31947): 	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3170)
E/AndroidRuntime(31947): 	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2846)
E/AndroidRuntime(31947): 	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3170)
E/AndroidRuntime(31947): 	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2846)
E/AndroidRuntime(31947): 	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3170)
E/AndroidRuntime(31947): 	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2846)
E/AndroidRuntime(31947): 	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3170)
E/AndroidRuntime(31947): 	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2846)
E/AndroidRuntime(31947): 	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3170)
E/AndroidRuntime(31947): 	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2846)
E/AndroidRuntime(31947): 	at com.android.internal.policy.DecorView.superDispatchTouchEvent(DecorView.java:569)
E/AndroidRuntime(31947): 	at com.android.internal.policy.PhoneWindow.superDispatchTouchEvent(PhoneWindow.java:1918)
E/AndroidRuntime(31947): 	at android.app.Dialog.dispatchTouchEvent(Dialog.java:960)
E/AndroidRuntime(31947): 	at android.inputmethodservice.SoftInputWindow.dispatchTouchEvent(SoftInputWindow.java:151)
E/AndroidRuntime(31947): 	at com.android.internal.policy.DecorView.dispatchTouchEvent(DecorView.java:517)
E/AndroidRuntime(31947): 	at android.view.View.dispatchPointerEvent(View.java:15384)
E/AndroidRuntime(31947): 	at android.view.ViewRootImpl$ViewPostImeInputStage.processPointerEvent(ViewRootImpl.java:6907)
E/AndroidRuntime(31947): 	at android.view.ViewRootImpl$ViewPostImeInputStage.onProcess(ViewRootImpl.java:6678)
E/AndroidRuntime(31947): 	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:6125)
E/AndroidRuntime(31947): 	at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:6187)
E/AndroidRuntime(31947): 	at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:6148)
E/AndroidRuntime(31947): 	at android.view.ViewRootImpl$AsyncInputStage.forward(ViewRootImpl.java:6322)
E/AndroidRuntime(31947): 	at android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:6156)
E/AndroidRuntime(31947): 	at android.view.ViewRootImpl$AsyncInputStage.apply(ViewRootImpl.java:6379)
E/AndroidRuntime(31947): 	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:6129)
E/AndroidRuntime(31947): 	at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:6187)
E/AndroidRuntime(31947): 	at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:6148)
E/AndroidRuntime(31947): 	at android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:6156)
E/AndroidRuntime(31947): 	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:6129)
E/AndroidRuntime(31947): 	at android.view.ViewRootImpl.deliverInputEvent(ViewRootImpl.java:9288)
E/AndroidRuntime(31947): 	at android.view.ViewRootImpl.doProcessInputEvents(ViewRootImpl.java:9239)
E/AndroidRuntime(31947): 	at android.view.ViewRootImpl.enqueueInputEvent(ViewRootImpl.java:9188)
E/AndroidRuntime(31947): 	at android.view.ViewRootImpl$WindowInputEventReceiver.onInputEvent(ViewRootImpl.java:9439)
E/AndroidRuntime(31947): 	at android.view.InputEventReceiver.dispatchInputEvent(InputEventReceiver.java:276)
E/AndroidRuntime(31947): 	at android.os.MessageQueue.nativePollOnce(Native Method)
E/AndroidRuntime(31947): 	at android.os.MessageQueue.next(MessageQueue.java:339)
E/AndroidRuntime(31947): 	at android.os.Looper.loopOnce(Looper.java:186)
E/AndroidRuntime(31947): 	at android.os.Looper.loop(Looper.java:351)
E/AndroidRuntime(31947): 	at android.app.ActivityThread.main(ActivityThread.java:8381)
E/AndroidRuntime(31947): 	at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime(31947): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:584)
E/AndroidRuntime(31947): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1013)
E/CustomActivityOnCrash(31947): App has crashed, executing CustomActivityOnCrash's UncaughtExceptionHandler
E/CustomActivityOnCrash(31947): java.lang.ArrayIndexOutOfBoundsException: length=33; index=-1
E/CustomActivityOnCrash(31947): 	at com.osfans.trime.ime.keyboard.KeyboardView$1.onFling(KeyboardView.java:527)
E/CustomActivityOnCrash(31947): 	at android.view.GestureDetector.onTouchEvent(GestureDetector.java:767)
E/CustomActivityOnCrash(31947): 	at com.osfans.trime.ime.keyboard.KeyboardView.onModifiedTouchEvent(KeyboardView.java:1580)
E/CustomActivityOnCrash(31947): 	at com.osfans.trime.ime.keyboard.KeyboardView.onTouchEvent(KeyboardView.java:1543)
E/CustomActivityOnCrash(31947): 	at android.view.View.dispatchTouchEvent(View.java:15111)
E/CustomActivityOnCrash(31947): 	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3170)
E/CustomActivityOnCrash(31947): 	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2846)
E/CustomActivityOnCrash(31947): 	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3170)
E/CustomActivityOnCrash(31947): 	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2846)
E/CustomActivityOnCrash(31947): 	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3170)
E/CustomActivityOnCrash(31947): 	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2846)
E/CustomActivityOnCrash(31947): 	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3170)
E/CustomActivityOnCrash(31947): 	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2846)
E/CustomActivityOnCrash(31947): 	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3170)
E/CustomActivityOnCrash(31947): 	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2846)
E/CustomActivityOnCrash(31947): 	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3170)
E/CustomActivityOnCrash(31947): 	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2846)
E/CustomActivityOnCrash(31947): 	at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3170)
E/CustomActivityOnCrash(31947): 	at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2846)
E/CustomActivityOnCrash(31947): 	at com.android.internal.policy.DecorView.superDispatchTouchEvent(DecorView.java:569)
E/CustomActivityOnCrash(31947): 	at com.android.internal.policy.PhoneWindow.superDispatchTouchEvent(PhoneWindow.java:1918)
E/CustomActivityOnCrash(31947): 	at android.app.Dialog.dispatchTouchEvent(Dialog.java:960)
E/CustomActivityOnCrash(31947): 	at android.inputmethodservice.SoftInputWindow.dispatchTouchEvent(SoftInputWindow.java:151)
E/CustomActivityOnCrash(31947): 	at com.android.internal.policy.DecorView.dispatchTouchEvent(DecorView.java:517)
E/CustomActivityOnCrash(31947): 	at android.view.View.dispatchPointerEvent(View.java:15384)
E/CustomActivityOnCrash(31947): 	at android.view.ViewRootImpl$ViewPostImeInputStage.processPointerEvent(ViewRootImpl.java:6907)
E/CustomActivityOnCrash(31947): 	at android.view.ViewRootImpl$ViewPostImeInputStage.onProcess(ViewRootImpl.java:6678)
E/CustomActivityOnCrash(31947): 	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:6125)
E/CustomActivityOnCrash(31947): 	at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:6187)
E/CustomActivityOnCrash(31947): 	at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:6148)
E/CustomActivityOnCrash(31947): 	at android.view.ViewRootImpl$AsyncInputStage.forward(ViewRootImpl.java:6322)
E/CustomActivityOnCrash(31947): 	at android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:6156)
E/CustomActivityOnCrash(31947): 	at android.view.ViewRootImpl$AsyncInputStage.apply(ViewRootImpl.java:6379)
E/CustomActivityOnCrash(31947): 	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:6129)
E/CustomActivityOnCrash(31947): 	at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:6187)
E/CustomActivityOnCrash(31947): 	at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:6148)
E/CustomActivityOnCrash(31947): 	at android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:6156)
E/CustomActivityOnCrash(31947): 	at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:6129)
E/CustomActivityOnCrash(31947): 	at android.view.ViewRootImpl.deliverInputEvent(ViewRootImpl.java:9288)
E/CustomActivityOnCrash(31947): 	at android.view.ViewRootImpl.doProcessInputEvents(ViewRootImpl.java:9239)
E/CustomActivityOnCrash(31947): 	at android.view.ViewRootImpl.enqueueInputEvent(ViewRootImpl.java:9188)
E/CustomActivityOnCrash(31947): 	at android.view.ViewRootImpl$WindowInputEventReceiver.onInputEvent(ViewRootImpl.java:9439)
E/CustomActivityOnCrash(31947): 	at android.view.InputEventReceiver.dispatchInputEvent(InputEventReceiver.java:276)
E/CustomActivityOnCrash(31947): 	at android.os.MessageQueue.nativePollOnce(Native Method)
E/CustomActivityOnCrash(31947): 	at android.os.MessageQueue.next(MessageQueue.java:339)
E/CustomActivityOnCrash(31947): 	at android.os.Looper.loopOnce(Looper.java:186)
E/CustomActivityOnCrash(31947): 	at android.os.Looper.loop(Looper.java:351)
E/CustomActivityOnCrash(31947): 	at android.app.ActivityThread.main(ActivityThread.java:8381)
E/CustomActivityOnCrash(31947): 	at java.lang.reflect.Method.invoke(Native Method)
E/CustomActivityOnCrash(31947): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:584)
E/CustomActivityOnCrash(31947): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1013)
E/CustomActivityOnCrash(31947): Failed when resolving the restart activity class via getLaunchIntentForPackage, stack trace follows!
E/CustomActivityOnCrash(31947): java.lang.ClassNotFoundException: com.osfans.trime.PrefLauncherAlias
E/CustomActivityOnCrash(31947): 	at java.lang.Class.classForName(Native Method)
E/CustomActivityOnCrash(31947): 	at java.lang.Class.forName(Class.java:454)
E/CustomActivityOnCrash(31947): 	at java.lang.Class.forName(Class.java:379)
E/CustomActivityOnCrash(31947): 	at cat.ereza.customactivityoncrash.CustomActivityOnCrash.getLauncherActivity(CustomActivityOnCrash.java:648)
E/CustomActivityOnCrash(31947): 	at cat.ereza.customactivityoncrash.CustomActivityOnCrash.guessRestartActivityClass(CustomActivityOnCrash.java:602)
E/CustomActivityOnCrash(31947): 	at cat.ereza.customactivityoncrash.CustomActivityOnCrash.lambda$install$0(CustomActivityOnCrash.java:177)
E/CustomActivityOnCrash(31947): 	at cat.ereza.customactivityoncrash.CustomActivityOnCrash$$ExternalSyntheticLambda0.uncaughtException(Unknown Source:2)
E/CustomActivityOnCrash(31947): 	at java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:1073)
E/CustomActivityOnCrash(31947): 	at java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:1068)
E/CustomActivityOnCrash(31947): 	at com.android.internal.os.RuntimeInitExtImpl.uncaughtExceptionExt(RuntimeInitExtImpl.java:44)
E/CustomActivityOnCrash(31947): 	at com.android.internal.os.RuntimeInit$LoggingHandler.uncaughtException(RuntimeInit.java:140)
E/CustomActivityOnCrash(31947): 	at java.lang.Thread.dispatchUncaughtException(Thread.java:2300)
E/CustomActivityOnCrash(31947): Caused by: java.lang.ClassNotFoundException: com.osfans.trime.PrefLauncherAlias
E/CustomActivityOnCrash(31947): 	... 12 more
I/Process (31947): Sending signal. PID: 31947 SIG: 9
```

**Screenshots**
If applicable, add screenshots to help explain your problem.

**Smartphone (please complete the following information):**
```
OS Name: CPH2415_11_C.26
OS Version: 5.10.101-android12-9-00013-g61478849d5b2-ab9281647 (S.e68b86-241-12da2)
OS API Level: 33
Device: OP5552L1
Model (product): CPH2415 (CPH2415)
Manufacturer: OnePlus
```
 - Build info: 
  
   commit: [56f1e2116b4a0abf0e19b53e8b9c56807a11b7c1](https://github.com/osfans/trime/tree/56f1e2116b4a0abf0e19b53e8b9c56807a11b7c1)
 ```   
Branch: develop
   Repository: https://github.com/osfans/trime
   ```
**Additional context**
Add any other context about the problem here.

--- a/app/src/main/java/com/osfans/trime/ime/core/TrimeInputMethodService.kt
+++ b/app/src/main/java/com/osfans/trime/ime/core/TrimeInputMethodService.kt
@@ -99,6 +99,9 @@ open class TrimeInputMethodService : LifecycleInputMethodService() {
     private var mCompositionPopupWindow: CompositionPopupWindow? = null
     var candidateExPage = false
 
+    var shouldUpdateRimeOption = false
+    var shouldResetAsciiMode = false
+
     private val cursorCapsMode: Int
         get() =
             currentInputEditorInfo.run {
@@ -175,17 +178,17 @@ open class TrimeInputMethodService : LifecycleInputMethodService() {
 
     fun loadConfig() {
         val theme = ThemeManager.activeTheme
-        textInputManager!!.shouldResetAsciiMode = theme.style.getBoolean("reset_ascii_mode")
+        shouldResetAsciiMode = theme.style.getBoolean("reset_ascii_mode")
         isAutoCaps = theme.style.getBoolean("auto_caps")
-        textInputManager!!.shouldUpdateRimeOption = true
+        shouldUpdateRimeOption = true
     }
 
     private fun updateRimeOption(): Boolean {
         try {
-            if (textInputManager!!.shouldUpdateRimeOption) {
+            if (shouldUpdateRimeOption) {
                 Rime.setOption("soft_cursor", prefs.keyboard.softCursorEnabled) // 軟光標
                 Rime.setOption("_horizontal", ThemeManager.activeTheme.style.getBoolean("horizontal")) // 水平模式
-                textInputManager!!.shouldUpdateRimeOption = false
+                shouldUpdateRimeOption = false
             }
         } catch (e: Exception) {
             e.printStackTrace()
@@ -326,11 +329,9 @@ open class TrimeInputMethodService : LifecycleInputMethodService() {
 
         loadConfig()
         KeyboardSwitcher.newOrReset()
-        if (textInputManager != null) {
-            textInputManager!!.shouldUpdateRimeOption = true // 不能在Rime.onMessage中調用set_option，會卡死
-            bindKeyboardToInputView()
-            updateComposing() // 切換主題時刷新候選
-        }
+        shouldUpdateRimeOption = true // 不能在Rime.onMessage中調用set_option，會卡死
+        bindKeyboardToInputView()
+        updateComposing() // 切換主題時刷新候選
         setInputView(inputView!!)
     }
 

--- a/app/src/main/java/com/osfans/trime/ime/keyboard/KeyboardView.kt
+++ b/app/src/main/java/com/osfans/trime/ime/keyboard/KeyboardView.kt
@@ -409,6 +409,7 @@ class KeyboardView
                     the sliding distance on the other coordinate axis.
                          */
                             if (mPossiblePoly) return false
+                            if (mDownKey == NOT_A_KEY) return false
                             val deltaX = me2.x - me1!!.x // distance X
                             val deltaY = me2.y - me1.y // distance Y
                             val absX = abs(deltaX) // absolute value of distance X

--- a/app/src/main/java/com/osfans/trime/ime/text/Composition.kt
+++ b/app/src/main/java/com/osfans/trime/ime/text/Composition.kt
@@ -470,7 +470,7 @@ class Composition(context: Context, attrs: AttributeSet?) : TextView(context, at
         ss!!.setSpan(getAlign(m), start, end, 0)
         movePos[0] = start
         movePos[1] = end
-        ss!!.setSpan(AbsoluteSizeSpan(keyTextSize), start, end, 0)
+        ss!!.setSpan(AbsoluteSizeSpan(sp(keyTextSize)), start, end, 0)
         ss!!.setSpan(ForegroundColorSpan(keyTextColor), start, end, 0)
         val suffix = CollectionUtils.obtainString(m, "end")
         if (suffix.isNotEmpty()) ss!!.append(suffix)

--- a/app/src/main/java/com/osfans/trime/ime/text/TextInputManager.kt
+++ b/app/src/main/java/com/osfans/trime/ime/text/TextInputManager.kt
@@ -63,9 +63,14 @@ class TextInputManager private constructor() :
         val locales = Array(2) { Locale.getDefault() }
 
         var needSendUpRimeKey: Boolean = false
-        var shouldUpdateRimeOption: Boolean = true
         var isComposable: Boolean = false
-        var shouldResetAsciiMode: Boolean = false
+
+        private var shouldUpdateRimeOption
+            get() = trime.shouldUpdateRimeOption
+            set(value) {
+                trime.shouldUpdateRimeOption = value
+            }
+        private val shouldResetAsciiMode get() = trime.shouldResetAsciiMode
 
         companion object {
             /** Delimiter regex for key property group, their format like `{property_1: value_1, property_2: value_2}` */

